File: docs/breez-sdk/snippets/swift/BreezSdkSnippets/Sources/Passkey.swift
import BreezSdkSpark
import Foundation

// ANCHOR: implement-prf-provider
// Implement PrfProvider for a custom authenticator (hardware key, FIDO2,
// file-backed). Only deriveSeeds and isSupported are required.
class CustomPrfProvider: PrfProvider {
    func deriveSeeds(request: DeriveSeedsRequest) async throws -> DeriveSeedsOutput {
        // Return one 32-byte PRF output per salt, in input order.
        fatalError("Implement using WebAuthn or native passkey APIs")
    }

    func isSupported() async throws -> Bool {
        fatalError("Check platform passkey availability")
    }

    func createPasskey(excludeCredentials: [Data]) async throws -> PasskeyCredential {
        // Register a credential and return its ID plus attestation.
        fatalError("Implement registration via WebAuthn create() / native API")
    }

    func checkDomainAssociation() async throws -> DomainAssociation {
        return .skipped(reason: "CustomPrfProvider does not verify domain association")
    }
}
// ANCHOR_END: implement-prf-provider

func checkAvailability() async throws {
    let prfProvider = PasskeyProvider(
        options: PasskeyProviderOptions(rpId: "<your-rp-domain>", rpName: "Your App")
    )
    let passkey = PasskeyClient(prfProvider: prfProvider, breezApiKey: "<breez api key>", config: nil)

    // ANCHOR: check-availability
    switch try await passkey.checkAvailability() {
    case .available:
        // Show passkey as primary option.
        break
    case .prfUnsupported:
        // Fall back to mnemonic flow.
        break
    case .notAssociated(let source, let reason):
        print("Domain association failed (source=\(source)): \(reason)")
    case .skipped:
        // No verification source on this platform; proceed normally.
        break
    }
    // ANCHOR_END: check-availability
}

func setupPasskeyClient() -> PasskeyClient {
    // ANCHOR: setup-client
    let passkey = PasskeyClient(
        breezApiKey: "<breez api key>",
        config: PasskeyConfig(
            providerOptions: PasskeyProviderOptions(rpId: "<your-rp-domain>", rpName: "Your App")
        )
    )
    // ANCHOR_END: setup-client
    return passkey
}

func connectWithPasskey() async throws -> BreezSdk {
    let prfProvider = PasskeyProvider(
        options: PasskeyProviderOptions(rpId: "<your-rp-domain>", rpName: "Your App")
    )
    let passkey = PasskeyClient(prfProvider: prfProvider, breezApiKey: "<breez api key>", config: nil)

    // ANCHOR: connect-with-passkey
    // Single-CTA onboarding: silent sign-in, fall through to register.
    var config = defaultConfig(network: .mainnet)
    config.apiKey = "<breez api key>"

    let response = try await passkey.connectWithPasskey(
        request: ConnectWithPasskeyRequest()
    )

    if response.labels.count > 1 {
        // Returning multi-wallet user: let them pick a label, then sign in to it.
        // let chosen = promptForLabel(response.labels)
        // return try await passkey.signIn(request: SignInRequest(label: chosen))
    }

    let sdk = try await connect(
        request: ConnectRequest(
            config: config,
            seed: response.wallet.seed,
            storageDir: "./.data"
        ))
    // ANCHOR_END: connect-with-passkey
    return sdk
}

func signInExistingUser() async throws -> SignInResponse {
    let prfProvider = PasskeyProvider(
        options: PasskeyProviderOptions(rpId: "<your-rp-domain>", rpName: "Your App")
    )
    let passkey = PasskeyClient(prfProvider: prfProvider, breezApiKey: "<breez api key>", config: nil)

    // ANCHOR: sign-in
    // Returning-user sign-in. No fall-through to register.
    return try await passkey.signIn(request: SignInRequest(label: "personal"))
    // ANCHOR_END: sign-in
}

func registerNewPasskey() async throws -> BreezSdk {
    let prfProvider = PasskeyProvider(
        options: PasskeyProviderOptions(rpId: "<your-rp-domain>", rpName: "Your App")
    )
    let passkey = PasskeyClient(prfProvider: prfProvider, breezApiKey: "<breez api key>", config: nil)

    // ANCHOR: register-passkey
    var config = defaultConfig(network: .mainnet)
    config.apiKey = "<breez api key>"

    let response = try await passkey.register(
        request: RegisterRequest(label: "personal")
    )

    let sdk = try await connect(
        request: ConnectRequest(
            config: config,
            seed: response.wallet.seed,
            storageDir: "./.data"
        ))
    // ANCHOR_END: register-passkey
    return sdk
}

func credentialMetadata() async throws {
    let prfProvider = PasskeyProvider(
        options: PasskeyProviderOptions(rpId: "<your-rp-domain>", rpName: "Your App")
    )
    let passkey = PasskeyClient(prfProvider: prfProvider, breezApiKey: nil, config: nil)

    // ANCHOR: credential-metadata
    let response = try await passkey.register(
        request: RegisterRequest(label: "personal")
    )

    if let credential = response.credential {
        // Persist to reopen the same wallet on sign-in
        print(credential.credentialId)
        // Authenticator model (display hint, unverified)
        print(credential.aaguid)
        // Whether the passkey syncs across devices
        print(credential.backupEligible)
    }

    // Pin the stored credential ID so the OS can't substitute a sibling
    // credential, which would derive a different wallet.
    let signInResponse = try await passkey.signIn(
        request: SignInRequest(
            label: "personal",
            allowCredentials: [
                // stored credentialId bytes
            ]
        )
    )
    // Pass to connect() to open the wallet
    print(signInResponse.wallet.seed)
    // Label this wallet was derived from
    print(signInResponse.wallet.label)
    // This passkey's labels (populated on discovery sign-in)
    print(signInResponse.labels)
    // Credential signed in with (credential_id only)
    print(signInResponse.credential)
    // ANCHOR_END: credential-metadata
}

func listLabels() async throws -> [String] {
    let prfProvider = PasskeyProvider(
        options: PasskeyProviderOptions(rpId: "<your-rp-domain>", rpName: "Your App")
    )
    let passkey = PasskeyClient(prfProvider: prfProvider, breezApiKey: "<breez api key>", config: nil)
    // ANCHOR: list-labels
    let labels = try await passkey.labels().list()
    for label in labels {
        print("Found label: \(label)")
    }
    // ANCHOR_END: list-labels
    return labels
}

func storeLabel() async throws {
    let prfProvider = PasskeyProvider(
        options: PasskeyProviderOptions(rpId: "<your-rp-domain>", rpName: "Your App")
    )
    let passkey = PasskeyClient(prfProvider: prfProvider, breezApiKey: "<breez api key>", config: nil)
    // ANCHOR: store-label
    try await passkey.labels().store(label: "personal")
    // ANCHOR_END: store-label
}

func checkDomain() async throws {
    // ANCHOR: domain-association
    // Lower-level provider call. Most hosts use `checkAvailability` instead.
    let prfProvider = PasskeyProvider(
        options: PasskeyProviderOptions(rpId: "<your-rp-domain>", rpName: "Your App")
    )
    let result = try await prfProvider.checkDomainAssociation()

    switch result {
    case .associated:
        break
    case .notAssociated(let source, let reason):
        print("Domain association failed (source=\(source)): \(reason)")
        return
    case .skipped:
        break
    }
    // ANCHOR_END: domain-association
}

func recoverFromAlreadyExists() async throws -> Wallet {
    let prfProvider = PasskeyProvider(
        options: PasskeyProviderOptions(rpId: "<your-rp-domain>", rpName: "Your App")
    )
    let passkey = PasskeyClient(prfProvider: prfProvider, breezApiKey: "<breez api key>", config: nil)

    // ANCHOR: recover-already-exists
    do {
        let response = try await passkey.register(
            request: RegisterRequest(
                label: "personal",
                excludeCredentials: [
                    // app-persisted credential IDs from prior registrations
                ]
            )
        )
        return response.wallet
    } catch PrfProviderError.CredentialAlreadyExists {
        // A matching credential already exists; sign in instead.
        let response = try await passkey.signIn(
            request: SignInRequest(label: "personal")
        )
        return response.wallet
    }
    // ANCHOR_END: recover-already-exists
}

func handleTimeout() async throws -> SignInResponse {
    let prfProvider = PasskeyProvider(
        options: PasskeyProviderOptions(rpId: "<your-rp-domain>", rpName: "Your App")
    )
    let passkey = PasskeyClient(prfProvider: prfProvider, breezApiKey: "<breez api key>", config: nil)

    // ANCHOR: handle-timeout
    do {
        return try await passkey.signIn(
            request: SignInRequest(label: "personal")
        )
    } catch PrfProviderError.UserTimedOut {
        // Show a retry UI. Do NOT auto-retry without user input.
        print("Sign-in timed out: show \"Try Again\" UI.")
        throw PrfProviderError.UserTimedOut
    }
    // ANCHOR_END: handle-timeout
}

File: docs/breez-sdk/snippets/kotlin_mpp_lib/shared/src/androidMain/kotlin/com/example/kotlinmpplib/Passkey.kt
package com.example.kotlinmpplib

import android.app.Activity
import breez_sdk_spark.*
import technology.breez.spark.passkey.PasskeyClient
import technology.breez.spark.passkey.PasskeyProvider

// ANCHOR: implement-prf-provider
// Implement PrfProvider for a custom authenticator. Only deriveSeeds and
// isSupported are required.
class CustomPrfProvider : PrfProvider {
    override suspend fun deriveSeeds(request: DeriveSeedsRequest): DeriveSeedsOutput {
        // Return one 32-byte PRF output per salt, in input order.
        TODO("Implement using WebAuthn or native passkey APIs")
    }

    override suspend fun isSupported(): Boolean {
        TODO("Check platform passkey availability")
    }

    override suspend fun createPasskey(excludeCredentials: List<ByteArray>): PasskeyCredential {
        // Register a credential and return its ID plus attestation.
        TODO("Implement registration via native passkey API")
    }

    override suspend fun checkDomainAssociation(): DomainAssociation {
        return DomainAssociation.Skipped("CustomPrfProvider does not verify domain association")
    }
}
// ANCHOR_END: implement-prf-provider

class PasskeySnippets(private val activity: Activity) {
    suspend fun checkAvailability() {
        val prfProvider = PasskeyProvider(
            activityProvider = { activity },
            options = PasskeyProviderOptions(rpId = "<your-rp-domain>", rpName = "Your App"),
        )
        val passkey = PasskeyClient(prfProvider, "<breez api key>", null)

        // ANCHOR: check-availability
        when (val availability = passkey.checkAvailability()) {
            is PasskeyAvailability.Available -> Unit
            is PasskeyAvailability.PrfUnsupported -> Unit
            is PasskeyAvailability.NotAssociated -> {
                // Log.e("Breez", "Domain association failed (source=${availability.source}): ${availability.reason}")
            }
            is PasskeyAvailability.Skipped -> Unit
        }
        // ANCHOR_END: check-availability
    }

    fun setupPasskeyClient(): PasskeyClient {
        // ANCHOR: setup-client
        val passkey = PasskeyClient(
            breezApiKey = "<breez api key>",
            activityProvider = { activity },
            config = PasskeyConfig(providerOptions = PasskeyProviderOptions(rpId = "<your-rp-domain>", rpName = "Your App")),
        )
        // ANCHOR_END: setup-client
        return passkey
    }

    suspend fun connectWithPasskey(): BreezSdk {
        val prfProvider = PasskeyProvider(
            activityProvider = { activity },
            options = PasskeyProviderOptions(rpId = "<your-rp-domain>", rpName = "Your App"),
        )
        val passkey = PasskeyClient(prfProvider, "<breez api key>", null)

        // ANCHOR: connect-with-passkey
        // Single-CTA onboarding: silent sign-in, fall through to register.
        val config = defaultConfig(Network.MAINNET).apply { apiKey = "<breez api key>" }
        val response = passkey.connectWithPasskey(
            ConnectWithPasskeyRequest(label = "personal")
        )

        val sdk = connect(ConnectRequest(config, response.wallet.seed, "./.data"))
        // ANCHOR_END: connect-with-passkey
        return sdk
    }

    suspend fun signInExistingUser(): SignInResponse {
        val prfProvider = PasskeyProvider(
            activityProvider = { activity },
            options = PasskeyProviderOptions(rpId = "<your-rp-domain>", rpName = "Your App"),
        )
        val passkey = PasskeyClient(prfProvider, "<breez api key>", null)

        // ANCHOR: sign-in
        // Returning-user sign-in. No fall-through to register.
        return passkey.signIn(SignInRequest(label = "personal"))
        // ANCHOR_END: sign-in
    }

    suspend fun registerNewPasskey(): BreezSdk {
        val prfProvider = PasskeyProvider(
            activityProvider = { activity },
            options = PasskeyProviderOptions(rpId = "<your-rp-domain>", rpName = "Your App"),
        )
        val passkey = PasskeyClient(prfProvider, "<breez api key>", null)

        // ANCHOR: register-passkey
        val config = defaultConfig(Network.MAINNET).apply { apiKey = "<breez api key>" }
        val response = passkey.register(RegisterRequest(label = "personal"))

        val sdk = connect(ConnectRequest(config, response.wallet.seed, "./.data"))
        // ANCHOR_END: register-passkey
        return sdk
    }

    suspend fun credentialMetadata() {
        val prfProvider = PasskeyProvider(
            activityProvider = { activity },
            options = PasskeyProviderOptions(rpId = "<your-rp-domain>", rpName = "Your App"),
        )
        val passkey = PasskeyClient(prfProvider, "<breez api key>", null)

        // ANCHOR: credential-metadata
        val response = passkey.register(RegisterRequest(label = "personal"))

        response.credential?.let { credential ->
            // Log.v("Breez", "${credential.credentialId}") // Persist to reopen the same wallet on sign-in
            // Log.v("Breez", "${credential.aaguid}") // Authenticator model (display hint, unverified)
            // Log.v("Breez", "${credential.backupEligible}") // Whether the passkey syncs across devices
        }

        // Pin the stored credential ID so the OS can't substitute a sibling
        // credential, which would derive a different wallet.
        val signInResponse = passkey.signIn(
            SignInRequest(
                label = "personal",
                allowCredentials = emptyList(), // stored credentialId bytes
            )
        )
        // Log.v("Breez", "${signInResponse.wallet.seed}") // Pass to connect() to open the wallet
        // Log.v("Breez", "${signInResponse.wallet.label}") // Label this wallet was derived from
        // Log.v("Breez", "${signInResponse.labels}") // This passkey's labels (populated on discovery sign-in)
        // Log.v("Breez", "${signInResponse.credential}") // Credential signed in with (credential_id only)
        // ANCHOR_END: credential-metadata
    }

    suspend fun listLabels(): List<String> {
        val prfProvider = PasskeyProvider(
            activityProvider = { activity },
            options = PasskeyProviderOptions(rpId = "<your-rp-domain>", rpName = "Your App"),
        )
        val passkey = PasskeyClient(prfProvider, "<breez api key>", null)
        // ANCHOR: list-labels
        val labels = passkey.labels().list()
        for (label in labels) {
            // Log.v("Breez", "Found label: $label")
        }
        // ANCHOR_END: list-labels
        return labels
    }

    suspend fun storeLabel() {
        val prfProvider = PasskeyProvider(
            activityProvider = { activity },
            options = PasskeyProviderOptions(rpId = "<your-rp-domain>", rpName = "Your App"),
        )
        val passkey = PasskeyClient(prfProvider, "<breez api key>", null)
        // ANCHOR: store-label
        passkey.labels().store("personal")
        // ANCHOR_END: store-label
    }

    suspend fun checkDomain() {
        // ANCHOR: domain-association
        // Diagnostic only: never blocks the ceremony.
        val prfProvider = PasskeyProvider(
            activityProvider = { activity },
            options = PasskeyProviderOptions(rpId = "<your-rp-domain>", rpName = "Your App"),
        )
        val result = prfProvider.checkDomainAssociation()

        when (result) {
            is DomainAssociation.Associated -> Unit
            is DomainAssociation.NotAssociated -> {
                // Log.e("Breez", "Domain association failed (source=${result.source}): ${result.reason}")
            }
            is DomainAssociation.Skipped -> Unit
        }
        // ANCHOR_END: domain-association
    }

    suspend fun recoverFromAlreadyExists(): Wallet {
        val prfProvider = PasskeyProvider(
            activityProvider = { activity },
            options = PasskeyProviderOptions(rpId = "<your-rp-domain>", rpName = "Your App"),
        )
        val passkey = PasskeyClient(prfProvider, "<breez api key>", null)

        // ANCHOR: recover-already-exists
        return try {
            val response = passkey.register(
                RegisterRequest(
                    label = "personal",
                    excludeCredentials = emptyList(), // app-persisted credential IDs from prior registrations
                )
            )
            response.wallet
        } catch (e: PrfProviderException.CredentialAlreadyExists) {
            // A matching credential already exists; sign in to it instead.
            val response = passkey.signIn(SignInRequest(label = "personal"))
            response.wallet
        }
        // ANCHOR_END: recover-already-exists
    }

    suspend fun handleTimeout(): SignInResponse {
        val prfProvider = PasskeyProvider(
            activityProvider = { activity },
            options = PasskeyProviderOptions(rpId = "<your-rp-domain>", rpName = "Your App"),
        )
        val passkey = PasskeyClient(prfProvider, "<breez api key>", null)

        // ANCHOR: handle-timeout
        return try {
            passkey.signIn(SignInRequest(label = "personal"))
        } catch (e: PrfProviderException.UserTimedOut) {
            // Show a retry UI. Do NOT auto-retry without user input.
            // Log.v("Breez", "Sign-in timed out: show \"Try Again\" UI.")
            throw e
        }
        // ANCHOR_END: handle-timeout
    }
}
